coins/monero/src/ringct/borromean.rs

```diff
@@ -0,0 +1,108 @@
+#![allow(non_snake_case)]
+
+use std::fmt::Debug;
+use std::io::{self, Read, Write};
+
+use curve25519_dalek::edwards::EdwardsPoint;
+use curve25519_dalek::scalar::Scalar;
+use curve25519_dalek::traits::Identity;
+
+use monero_generators::H2;
+
+use crate::hash_to_scalar;
+use crate::serialize::*;
+
+/// A Borromean signature.
+///
+/// Note: This type keeps the data as raw bytes as Monero has
+/// some transactions with unreduced scalars in this field, we
+/// could use `from_bytes_mod_order` but then we would not be able
+/// to encode this back into it's original form.
+///
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct BorroSig {
+  pub s0: [[u8; 32]; 64],
+  pub s1: [[u8; 32]; 64],
+  pub ee: [u8; 32],
+}
+
+impl BorroSig {
+  pub fn read<R: Read>(r: &mut R) -> io::Result<BorroSig> {
+    Ok(BorroSig {
+      s0: read_array(read_bytes, r)?,
+      s1: read_array(read_bytes, r)?,
+      ee: read_bytes(r)?,
+    })
+  }
+  pub fn write<W: Write>(&self, w: &mut W) -> io::Result<()> {
+    for s0 in self.s0.iter() {
+      w.write_all(s0)?;
+    }
+    for s1 in self.s1.iter() {
+      w.write_all(s1)?;
+    }
+    w.write_all(&self.ee)
+  }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct RangeSig {
+  pub asig: BorroSig,
+  pub Ci: [EdwardsPoint; 64],
+}
+
+impl RangeSig {
+  pub fn read<R: Read>(r: &mut R) -> io::Result<RangeSig> {
+    Ok(RangeSig { asig: BorroSig::read(r)?, Ci: read_array(read_point, r)? })
+  }
+  pub fn write<W: Write>(&self, w: &mut W) -> io::Result<()> {
+    self.asig.write(w)?;
+    write_raw_vec(write_point, &self.Ci, w)
+  }
+
+  pub fn verify(&self, commitment: &EdwardsPoint) -> bool {
+    let mut P1 = Vec::with_capacity(64);
+    let mut P2 = Vec::with_capacity(64);
+    let mut bbs0 = Vec::with_capacity(64);
+    let mut bbs1 = Vec::with_capacity(64);
+
+    let bbee = Scalar::from_bytes_mod_order(self.asig.ee);
+
+    let mut C_temp = EdwardsPoint::identity();
+
+    for i in 0 .. 64 {
+      bbs0.push(Scalar::from_bytes_mod_order(self.asig.s0[i]));
+      bbs1.push(Scalar::from_bytes_mod_order(self.asig.s1[i]));
+
+      P1.push(self.Ci[i]);
+      P2.push(P1[i] - H2[i]);
+
+      C_temp += P1[i];
+    }
+
+    if &C_temp != commitment {
+      false
+    } else {
+      verify_borromean(P1, P2, bbee, bbs0, bbs1)
```

Comment:
This is RangeSig::verify. verify_borromean just verifies each bit of the amount is a 1 or 0 whereas this also checks that Ci bits represent C

Reply:
Right, but this isn't a RangeSig because a RangeSig doesn't exist. These are Borromean ring signatures used to prove a value is in range. You can delineate Borromean, the academic, out from its usage (keeping the indepent function). I actually would like that. But in that case, you should ideally re-define this range proof as a Vec of Borromean signatures, instead of inlining the signatures' elements.